pkg/prowgen/podspec.go

```diff
@@ -0,0 +1,412 @@
+package prowgen
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	cioperatorapi "github.com/openshift/ci-tools/pkg/api"
+)
+
+var defaultPodSpec = corev1.PodSpec{
+	ServiceAccountName: "ci-operator",
+	Containers: []corev1.Container{
+		{
+			Args: []string{
+				"--image-import-pull-secret=/etc/pull-secret/.dockerconfigjson",
+				"--gcs-upload-secret=/secrets/gcs/service-account.json",
+				"--report-credentials-file=/etc/report/credentials",
+			},
+			Command:         []string{"ci-operator"},
+			Image:           "ci-operator:latest",
+			ImagePullPolicy: corev1.PullAlways,
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{"cpu": *resource.NewMilliQuantity(10, resource.DecimalSI)},
+			},
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      "pull-secret",
+					MountPath: "/etc/pull-secret",
+					ReadOnly:  true,
+				},
+				{
+					Name:      "result-aggregator",
+					MountPath: "/etc/report",
+					ReadOnly:  true,
+				},
+				{
+					Name:      "gcs-credentials",
+					MountPath: cioperatorapi.GCSUploadCredentialsSecretMountPath,
+					ReadOnly:  true,
+				},
+			},
+		},
+	},
+	Volumes: []corev1.Volume{
+		{
+			Name: "pull-secret",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{SecretName: "registry-pull-credentials"},
+			},
+		},
+		{
+			Name: "result-aggregator",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{SecretName: "result-aggregator"},
+			},
+		},
+	},
+}
+
+type ciOperatorPodSpecGenerator struct {
+	meta        cioperatorapi.Metadata
+	buildErrors []error
+
+	needsBoskos     bool
+	needsPullSecret bool
+
+	claims                           bool
+	promotion                        bool
+	ghToken                          bool
+	reuseGitHubTokenDecorationVolume bool
+
+	releaseRpmsVersion string
+	clusterProfile     cioperatorapi.ClusterProfile
+	template           string
+	templateCommand    string
+	templateTestIST    string
+
+	targets sets.String
+	secrets []cioperatorapi.Secret
+}
+
+// CiOperatorPodSpecGenerator is a builder-pattern for building PodSpecs that
+// run ci-operator for various purposes, ensuring the consistency of volumes,
+// mounts, parameters and other content.
+type CiOperatorPodSpecGenerator interface {
```

Comment:
This is backwards from the normal builder pattern, right? Each new functionality must be added to this mega-interface and as fields to the data struct. Instead, couldn't we do something like

```go
type mutator func(*v1.PodSpec) error

type ciOperatorPodSpecGenerator struct {
    meta cioperatorapi.Metadata
    mutators []mutator
}

func NewCIOperatorPodSpecGenerator(meta cioperatorapi.Metadata, mutators ...mutator) *ciOperatorPodSpecGenerator {
    // ...
}

func CIPullSecret() mutator {
    return func(spec *v1.PodSpec) error {
        handleSecret(spec, cioperatorapi.Secret{Name: "ci-pull-credentials"})
        return nil
    }
}
```

Then the application is encapsulated to any individual thing, as long as they're independent?

Reply:
(you already have a good encapsulation of intent in the `handle*` functions, but unless we need complex multi-option dependencies we should be able to get away without a mega-struct of options?)

Reply:
Hmm...  Interesting, I know *this* to be a normal builder (I agree it's OOPy) - the job level plays the role of a director, you accumulate internal state and eventually build the whole thing.

I like your pattern and the suggestion to get rid of the interface and internal state, but I think some of the things needed here would be hard if we simply iterated mutators over shared `PodSpec` state. There ** are** some dependencies between individual mutators (mostly related to templates: templates need cluster profiles, must have one target, some need RPMs, but RPMs do not make sense outside of templates...), plus accumulating simple intents makes it easier to be idempotent (e.g. to avoid mistakenly add one secret or target twice). It's also easier to do some sanity checking to error out on the final `PodSpec` - this is easier to do over intents than over the final product.

I originally had something like you propose - the builder operated over an internal PodSpec and eventually just returned it, but it was really annoying to work with internally.

Reply:
I wonder if I could simplify the thing at least partially though... Parts of the internal state could be replaced with easy non-state mutators like you suggest. The interface could be generalized to accept something abstract...

Reply:
OK, if we have complicated inter-mutator dependencies (ew) then we can't do anything super simple like I had in mind. Let me know if you decide to change anything or if you want to just review as-is

Reply:
They are not that complicated, but there are deps. I have a partial change to make this PR simpler, I hope to update the PR today.